internal/requirements/requirements.go

```diff
@@ -0,0 +1,157 @@
+package requirements
+
+import (
+	"bufio"
+	"bytes"
+	"io"
+	"strings"
+
+	"golang.org/x/text/encoding"
+	"golang.org/x/text/encoding/unicode"
+)
+
+type requirementsTxt struct {
+	encoder *encoding.Encoder // encoder for reencoding requiremnets.txt
+	lines   []string          // decoded requirements
+	crlf    bool              // detected crlf line termination
+}
+
+func Read(r io.Reader) (*requirementsTxt, error) {
+	data, err := io.ReadAll(r)
+	if err != nil {
+		return nil, err
+	}
+
+	encoding := unicode.UTF8
+	for _, bom := range boms {
+		if bom.Detect(data) {
+			encoding = bom.Encoding()
+			break
+		}
+	}
+
+	decoder := encoding.NewDecoder()
+	data, err = decoder.Bytes(data)
+	if err != nil {
+		return nil, err
+	}
+
+	lines, crlf, err := readLines(data)
+	if err != nil {
+		return nil, err
+	}
+
+	return &requirementsTxt{
+		encoder: encoding.NewEncoder(),
+		lines:   lines,
+		crlf:    crlf,
+	}, nil
+}
+
+func (r *requirementsTxt) Write(w io.Writer) error {
+	ew := r.encoder.Writer(w)
+
+	lineEnding := []byte("\n")
+	if r.crlf {
+		lineEnding = []byte("\r\n")
+	}
+
+	for i, l := range r.lines {
+		if i > 0 {
+			if _, err := ew.Write(lineEnding); err != nil {
+				return err
+			}
+		}
+
+		if _, err := ew.Write([]byte(l)); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *requirementsTxt) Add(added string) {
+	for _, l := range r.lines {
+		if strings.Contains(l, added) {
+			return
+		}
+	}
+
+	lastLine := r.lines[len(r.lines)-1]
+	if lastLine == "" {
+		// if the last line is empty, insert above the empty line
+		r.lines[len(r.lines)-1] = added
+		r.lines = append(r.lines, "")
+	} else {
+		r.lines = append(r.lines, added)
+		r.lines = append(r.lines, "")
+	}
+}
+
+// dropCR drops a terminal \r from the data, and return true.
+func dropCR(data []byte) ([]byte, bool) {
+	if len(data) > 0 && data[len(data)-1] == '\r' {
```

Comment:
Maybe `if bytes.HasSuffix(data, []byte("\r")) {` would be readable.

Reply:
True. This is basically just copied from the built-in version.

Reply:
I changed it to your suggestion, except using `[]byte{'\r'}`